### PR TITLE
HDDS-4265. Refactor OzoneQuota to make it easy to support more quota …

### PR DESCRIPTION
**Pull request by wycccccc**
## What changes were proposed in this pull request?

This task is to make it easy to support more quota type.
I replaced the repeated judgment statements, and now just configure the new type 'setQuotaList' to complete the addition.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4265

## How was this patch tested?

Added unit test.

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 1]
```diff
@@ -43,13 +43,25 @@
...
-  public enum Units {UNDEFINED, BYTES, KB, MB, GB, TB}
+  public enum Units {BYTES, KB, MB, GB, TB}
 
   // Quota to decide how many buckets can be created.
   private long quotaInCounts;
   // Quota to decide how many storage space will be used in bytes.
   private long quotaInBytes;
   private RawQuotaInBytes rawQuotaInBytes;
+  // Data class of Quota.
+  private static QuotaList quotaList;
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:54]: we can Initialize quotaList in static {} 
```suggestion
  private static QuotaList quotaList;
  static {
     quotaList = new QuotaList();
     quotaList.addQuotaList(OZONE_QUOTA_TB, Units.TB, TB);
     quotaList.addQuotaList(OZONE_QUOTA_GB, Units.GB, GB);
     quotaList.addQuotaList(OZONE_QUOTA_MB, Units.MB, MB);
     quotaList.addQuotaList(OZONE_QUOTA_KB, Units.KB, KB);
     quotaList.addQuotaList(OZONE_QUOTA_BYTES, Units.BYTES, 1L);
  }
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 2]
```diff
@@ -72,24 +84,18 @@ public long getSize() {
...
-      case GB:
-        return this.getSize() * GB;
-      case TB:
-        return this.getSize() * TB;
-      case UNDEFINED:
-      default:
-        return -1;
+      long sQuote = -1L;
+      setQuotaList();
+      for(Units quota:quotaList.getUnitQuotaArray()){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:92]: space
```suggestion
      for(Units quota : quotaList.getUnitQuotaArray()){
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 3]
```diff
@@ -164,45 +171,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-        currUnit = Units.TB;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), TB);
-      }
-
-      if (uppercase.endsWith(OZONE_QUOTA_BYTES)) {
-        size = uppercase
-            .substring(0, uppercase.length() - OZONE_QUOTA_BYTES.length());
-        currUnit = Units.BYTES;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), 1L);
+      for(String quota:quotaList.getOzoneQuotaArray()){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:174]: space

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 4]
```diff
@@ -227,24 +210,15 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-      size = quotaInBytes / KB;
-      unit = Units.KB;
-    } else {
-      size = quotaInBytes;
-      unit = Units.BYTES;
+                                         long quotaInCounts) {
+    long size = 1L;
+    Units unit = Units.BYTES;
+    setQuotaList();
+    for (Long quota:quotaList.getSizeQuotaArray()){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:217]: space

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 5]
```diff
@@ -227,24 +210,15 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-    } else if (quotaInBytes % MB == 0) {
-      size = quotaInBytes / MB;
-      unit = Units.MB;
-    } else if (quotaInBytes % KB == 0) {
-      size = quotaInBytes / KB;
-      unit = Units.KB;
-    } else {
-      size = quotaInBytes;
-      unit = Units.BYTES;
+                                         long quotaInCounts) {
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:213]: indentation

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 6]
```diff
@@ -164,45 +171,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
+          break;
+        }
       }
       nSize = Long.parseLong(size);
     } catch (NumberFormatException e) {
       throw new IllegalArgumentException("Invalid values for quota, to ensure" +
-          " that the Quota format is legal(supported values are BYTES, KB, " +
-          "MB, GB and TB).");
+              " that the Quota format is legal(supported values are BYTES, " +
+              " KB, MB, GB and TB).");
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:188]: indentation

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 7]
```diff
@@ -72,24 +84,18 @@ public long getSize() {
...
-      case MB:
-        return this.getSize() * MB;
-      case GB:
-        return this.getSize() * GB;
-      case TB:
-        return this.getSize() * TB;
-      case UNDEFINED:
-      default:
-        return -1;
+      long sQuote = -1L;
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:90]: Should it be sQuota?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 8]
```diff
@@ -164,45 +171,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-            .substring(0, uppercase.length() - OZONE_QUOTA_BYTES.length());
-        currUnit = Units.BYTES;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), 1L);
+      for(String quota:quotaList.getOzoneQuotaArray()){
+        if (uppercase.endsWith((quota))){
+          size = uppercase
+                  .substring(0, uppercase.length() - quota.length());
+          currUnit = quotaList.getUnits(quota);
+          quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size),
+                  quotaList.getQuotaSize(currUnit));
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:180]: indentation

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 9]
```diff
@@ -164,45 +171,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-
-      if (uppercase.endsWith(OZONE_QUOTA_BYTES)) {
-        size = uppercase
-            .substring(0, uppercase.length() - OZONE_QUOTA_BYTES.length());
-        currUnit = Units.BYTES;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), 1L);
+      for(String quota:quotaList.getOzoneQuotaArray()){
+        if (uppercase.endsWith((quota))){
+          size = uppercase
+                  .substring(0, uppercase.length() - quota.length());
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:177]: indentation

**Review by captainzmc — COMMENTED** [review]
Thank @wycccccc's patch, there are some comments.

**Comment by wycccccc** [comment]
Thanks for your  review @captainzmc , I have resolved these issues.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java (hunk 4 above)
Comment (captainzmc) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:214]:
```suggestion
    for (Long quota : quotaList.getSizeQuotaArray()) {
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 10]
```diff
@@ -228,23 +209,13 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-    } else if (quotaInBytes % KB == 0) {
-      size = quotaInBytes / KB;
-      unit = Units.KB;
-    } else {
-      size = quotaInBytes;
-      unit = Units.BYTES;
+    long size = 1L;
+    Units unit = Units.BYTES;
+    for (Long quota:quotaList.getSizeQuotaArray()){
+      if(quotaInBytes % quota == 0){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:215]:
```suggestion
      if (quotaInBytes % quota == 0) {
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 11]
```diff
@@ -72,24 +84,17 @@ public long getSize() {
...
-        return this.getSize() * MB;
-      case GB:
-        return this.getSize() * GB;
-      case TB:
-        return this.getSize() * TB;
-      case UNDEFINED:
-      default:
-        return -1;
+      long sQuota = -1L;
+      for(Units quota : quotaList.getUnitQuotaArray()){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:91]:
```suggestion
      for (Units quota : quotaList.getUnitQuotaArray()) {
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 12]
```diff
@@ -72,24 +84,17 @@ public long getSize() {
...
+        if(quota == this.unit){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:92]:
```suggestion
        if (quota == this.unit) {
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 13]
```diff
@@ -164,45 +169,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
-        currUnit = Units.TB;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), TB);
-      }
-
-      if (uppercase.endsWith(OZONE_QUOTA_BYTES)) {
-        size = uppercase
-            .substring(0, uppercase.length() - OZONE_QUOTA_BYTES.length());
-        currUnit = Units.BYTES;
-        quotaMultiplyExact = Math.multiplyExact(Long.parseLong(size), 1L);
+      for(String quota : quotaList.getOzoneQuotaArray()){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:172]:
```suggestion
      for (String quota : quotaList.getOzoneQuotaArray()) {
```

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java [hunk 14]
```diff
@@ -164,45 +169,21 @@ public static OzoneQuota parseQuota(String quotaInBytes,
...
+        if (uppercase.endsWith((quota))){
```
Comment (captainzmc) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/OzoneQuota.java:173]:
```suggestion
        if (uppercase.endsWith((quota))) {
```

**Review by captainzmc — COMMENTED** [review]
Thanks for the update. LGTM, just few minor style issues.

**Comment by wycccccc** [comment]
Thanks for checking again@captainzmc, I have fixed them.

**Review by captainzmc — APPROVED** [review]
+1, LGTM.  After the CI executes successfully, let's merge this.

**Comment by captainzmc** [comment]
Merged this PR. Thanks for @wycccccc‘s patch.

**Pull request merged**